### PR TITLE
MTTS: Yonaguni/Europa version.

### PR DESCRIPTION
**Pull request by tigercat2000**
This commit adds a system known as "Modular Train/Tram System".
This system allows for modular moving trams, which will carry anything on
top of them. Atmospherics not included, as the system is based entirely on
objects instead of turfs.
@Nienhaus 
@Zuhayr

### CODE REVIEW

**Comment by Nienhaus** [comment]
;p

**Comment by tigercat2000** [comment]
I hope I targeted the right branch, since no one answered my question.

**Comment by Nienhaus** [comment]
Oh sorry, I have no idea. I don't think it matters at this point.

**Comment by GinjaNinja32** [comment]
>For now everything can just go to the master branch and we'll shake bugs out as we go, but in the future we'll need probably to sit down and sort out a dev cycle of some kind.
http://prevess.com/water/viewtopic.php?f=5&t=3

**Comment by tigercat2000** [comment]
Ah, wonderful.

**Comment by Zuhayr** [comment]
How stable/buggy is the code? Paradise was kinda vague on feedback, I'd very much like to use these as an intrinsic part of the map if they aren't tweaky.

**Comment by tigercat2000** [comment]
@Zuhayr It hasn't undergone any significant stress-testing, but it's fairly stable. It gets a bit odd if anything gets in the way of it's tracks, but I've never seen it actually get stuck- it just reverses it's direction.

**Comment by Nienhaus** [comment]
Can it be made to run over people?

**Comment by tigercat2000** [comment]
@Nienhaus not easily, since anything on the same turf as the tram that is anchored will be picked up and become part of the tram

**Comment by mwerezak** [comment]
You could maintain a list of things that are supposed to be on the tram and run over anything that is not in that list.

**Comment by Nienhaus** [comment]
:+1:

**Comment by tigercat2000** [comment]
@mwerezak How do you intend to differentiate between a mob that walked onto it at a station versus one it bumps into? It keeps a list of what is supposed to be on it, which updates every process, so it can't leave any mobs behind.

**Comment by Dennok** [comment]
Yes :+1:

**Comment by Dennok** [comment]
all trams can have list of all what in it.

**Comment by Dennok** [comment]
Also i make some similar to this and can make gen_collision() some faster for big trains.

**Comment by mwerezak** [comment]
@tigercat2000 Using a list...
> You could maintain a list of things that are supposed to be on the tram and run over anything that is not in that list.

**Comment by Dennok** [comment]
Need list what below. Its better.

**Comment by mwerezak** [comment]
@tigercat2000 I was thinking if the tram tracks whether it is or is not in motion, then when it updates it's list of contents it could check if the atom was in the old list.

**Comment by tigercat2000** [comment]
@mwerezak I really don't see a way to get it to work like that, since the trams are running their movement as long as they are turned on.

**Comment by mwerezak** [comment]
They stop at platforms don't they? How does a tram know when it's supposed to be moving around versus when it is stationary?

**Comment by tigercat2000** [comment]
@mwerezak it looks at the rail, determines if it has a stop_duration defined, then just refuses to listen to the loop after adding the stop_duration to world.time and waiting for world.time to catch up

**Comment by Dennok** [comment]
also need cut interactions with what below

File: code/modules/tram/tram.dm [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+/obj/tram/tram_controller/proc/init_tram()
+	tram = tram_floors + tram_walls //Tram is everything that makes up the tram
+	for(var/obj/tram/OT in tram)
+		var/turf/T = get_turf(OT)
+		for(var/atom/movable/AM in T) //Including anything inside of it
+			if(AM in tram)	continue
+			if(!check_validity(AM))	continue
+			tram += AM
+			if(!(src in tram))
+				tram += src
```
Comment (mwerezak) [inline code/modules/tram/tram.dm]: Is this supposed to be inside both of these loops?

Comment (tigercat2000) [inline code/modules/tram/tram.dm]: probably not, but it works

**Comment by Dennok** [comment]
@tigercat2000 Actually it move in one dir?

**Comment by tigercat2000** [comment]
@Dennok I'm not sure what you mean.

**Comment by mwerezak** [comment]
Very rough, but hopefully gets the idea across
```dm
/obj/tram/tram_controller
	var/list/tram_parts

/obj/tram/tram_controller/New()
	//...
	init_floors() //Search and link floors
	init_walls() //Search and link walls
	tram_parts = tram_floors + tram_walls
	tram_parts += src
	//...

/obj/tram/tram_controller/proc/update_tram()
	var/list/oldtram = tram
	
	tram = list()
	var/is_moving = is_moving()
	for(var/atom/movable/AM in get_passengers())
		if(is_moving && !(AM in oldtram))
			squish(AM)
		if(check_validity(AM)) //probably should exclude squished atoms or something
			tram += AM

//Gets everything in the tram that is not the tram
/obj/tram/tram_controller/proc/get_passengers()
	. = list()
	for(var/obj/tram/OT in tram)
		var/turf/T = get_turf(OT)
		for(var/atom/movable/AM in T) //Including anything inside of it
			if(AM in tram_parts) continue
			tram += AM

/obj/tram/tram_controller/proc/init_tram()
	tram = tram_floors + tram_walls //Tram is everything that makes up the tram
	for(var/atom/movable/AM in get_passengers())
		if(check_validity(AM))
			tram += AM

/obj/tram/tram_controller/proc/is_moving()
/obj/tram/tram_controller/proc/squish(var/atom/movable/AM)
```

**Comment by tigercat2000** [comment]
@mwerezak That does make a bit of sense, I'll see what I can do. No promises.

**Comment by Dennok** [comment]
Also it cant rotate, yet.

**Comment by tigercat2000** [comment]
@Dennok I couldn't think of any way to calculate the positioning required to rotate everything in the tram.

File: code/modules/tram/tram.dm (hunk 1 above)
Comment (mwerezak) [inline code/modules/tram/tram.dm]: It means for every AM you will also have a src, doubling the number of items that are not tram_floors or tram_walls...

**Comment by Vivalas** [comment]
I'm working on a system that works and allows the train to run over people and damage dense objects. It just takes mobs it finds in gen_collision, I don't see the big issue here, gen_collision doesn't count for mobs inside the tram

**Comment by Vivalas** [comment]
gen_collision looks at all the turfs in front of the tram before the tram moves, to determine if the tram should move in the first place, I have been able to make it run over mobs and destroy objects pretty easily

**Comment by tigercat2000** [comment]
gen_collision looks at every turf near the tram to determine if it would collide if it is trying to go X direction, it doesn't actually know which way the tram is going.

**Comment by Vivalas** [comment]
I've added a "crash list" that items found by gen_collision get added to with the index of cdir, then, in handle_collision it checks the crash list for all objects with the cdir of the direction the tram is moving, and handles it appropriately

**Pull request merged**